internal/oapispec/ffi2swagger.go

```diff
@@ -35,35 +38,101 @@ func NewFFISwaggerGen() FFISwaggerGen {
 	return &ffiSwaggerGen{}
 }
 
-func (og *ffiSwaggerGen) Generate(ctx context.Context, baseURL string, ffi *fftypes.FFI) (swagger *openapi3.T, err error) {
-
-	// routes := []*Route{}
-
-	// for _, method := range ffi.Methods {
-	// 	if routes, err = og.addMethod(ctx, routes, method); err != nil {
-	// 		return nil, err
-	// 	}
-	// }
-
-	// for _, event := range ffi.Events {
-	// 	if routes, err = og.addEvent(ctx, routes, event); err != nil {
-	// 		return nil, err
-	// 	}
-	// }
-
-	return nil, nil
+func (og *ffiSwaggerGen) Generate(ctx context.Context, baseURL string, api *fftypes.ContractAPI, ffi *fftypes.FFI) (swagger *openapi3.T, err error) {
+	hasLocation := api.Location.String() != "null"
+
+	routes := []*Route{}
+	for _, method := range ffi.Methods {
+		routes = og.addMethod(routes, method, hasLocation)
+	}
+	for _, event := range ffi.Events {
+		routes = og.addEvent(routes, event, hasLocation)
+	}
+
+	return SwaggerGen(ctx, routes, &SwaggerGenConfig{
+		Title:       ffi.Name,
+		Version:     ffi.Version,
+		Description: ffi.Description,
+		BaseURL:     baseURL,
+	}), nil
 }
 
-// func (og *ffiSwaggerGen) addMethod(ctx context.Context, routes []*Route, method *fftypes.FFIMethod) ([]*Route, error) {
+func (og *ffiSwaggerGen) addMethod(routes []*Route, method *fftypes.FFIMethod, hasLocation bool) []*Route {
+	return append(routes, &Route{
+		Name:             fmt.Sprintf("invoke_%s", method.Pathname),
+		Path:             fmt.Sprintf("invoke/%s", method.Pathname),
+		Method:           http.MethodPost,
+		JSONInputValue:   func() interface{} { return &fftypes.InvokeContractRequest{} },
+		JSONInputSchema:  func(ctx context.Context) string { return invokeRequestJSON(&method.Params, hasLocation).String() },
```

Comment:
Trying to understand this one.... this seems to be generating a `fftypes.JSONObject` - but won't this mean the schema just gets shown as `{}`?

Shouldn't it just be a `&fftypes.InvokeContractRequest{}` that gets returned?
... or even better have separate `&fftypes.InvokeContractRequestWithLocation{}` and `&fftypes.InvokeContractRequest{}` types?

Reply:
Yes, it ultimately is a `&fftypes.InvokeContractRequest{}` - but that struct has an opaque map of strings for its `Params` field. We generate the input schema by hand as a JSON string so that we can fill in the actual param names and types within that map.

And yes, I am building `JSONObject` for convenience and converting it to string. Technically it would be possible to build the string schema directly, but it seemed messy.